core/assertion.ts

```diff
@@ -66,7 +66,13 @@ export class Assertion {
   }
 
   public schema(schema: string) {
-    this.proto.target = this.session.target(schema, this.session.config.assertionSchema);
+    if (this.session.findActions({ schema: schema, name: this.proto.target.name }).length === 0) {
```

Comment:
partly to prevent code duplication, and partly because we now have a bunch of different circumstances where action schemas/names can be updated fairly dynamically prior to compile time, i think we should move these checks out of each Action, plus calling `checkTargetIsUnused` inside `session.ts` into a single place: compile time. I think we should run `checkTargetIsUnused` inside `session.compileGraphChunk`, and only in there.

Reply:
I'm a bit confused here. Earlier (https://github.com/dataform-co/dataform/pull/352#discussion_r314660766) you suggested to put `checkTargetIsUnused` inside `setNameAndTarget` and that is where it is located now.
I anyway agree with you that it would be better if that check happened just once and inside `compileGraphChunk` because that one is run for each and every action so the code would look cleaner and easier to maintain.

I actually tried to remove it from `setNameAndTarget` and put it only inside `compileGraphChunk` but it did not work (many of the duplicate action tests failed). Here is the code which I tested:

````
  public compileGraphChunk<T>(actions: Array<{ proto: IActionProto; compile(): T }>): T[] {
    const compiledChunks: T[] = [];

    actions.forEach(action => {
      try {
        if (!(action instanceof test.Test)) {
          this.checkTargetIsUnused(action.proto.target); // <---- THIS DID NOT WORK
        }
        const compiledChunk = action.compile();
        compiledChunks.push(compiledChunk);
      } catch (e) {
        this.compileError(e, action.proto.fileName);
      }
    });

    return compiledChunks;
  }
````